File: aery_fix.py
import os
from typing import List, Tuple

from bms_fs import (
    REPLACE_OPTION_UPDATE_PACK,
    move_elements_across_dir,
    bms_dir_similarity,
)


def main():
    src_dir = input("Input the src dir:")
    if src_dir.startswith('"') and src_dir.endswith('"'):
        src_dir = src_dir[1:-1]
    if not os.path.isdir(src_dir):
        print(f"{src_dir}: not a dir.")
        return
    dirs = [
        dir for dir in os.listdir(src_dir) if os.path.isdir(os.path.join(src_dir, dir))
    ]
    # 扫描所有带aery的目录
    aery_pair: List[Tuple[str, str, float]] = []
    for dir in dirs:
        if dir.find("Aery") == -1 and dir.find("AERY") == -1:
            continue
        # 源目录
        dir_path = os.path.join(src_dir, dir)
        for i in range(len(dir)):
            # 按长度寻找，排除自身，确认只剩下一个结果
            sub_len = i + 1
            scan_dirs = [
                sub_dir
                for sub_dir in dirs
                if sub_dir.startswith(dir[:sub_len]) and sub_dir != dir
            ]
            if len(scan_dirs) != 1:
                continue
            scan_dir = scan_dirs[0]
            # 目标路径
            scan_dir_path = os.path.join(src_dir, scan_dir)
            # 参数
            similarity = bms_dir_similarity(dir_path, scan_dir_path)
            aery_pair.append((dir_path, scan_dir_path, similarity))
            break
    # 打印待移动的部分
    for p in aery_pair:
        print(p)

    similarity_border = 0.95
    confirm = input(f"Confirm? (border: {similarity_border}) [y/N]")
    if not confirm.lower().startswith("y"):
        return

    for p in aery_pair:
        p_from: str = p[0]
        p_to: str = p[1]
        p_similarity: float = p[2]
        if p_similarity < similarity_border:
            continue
        print(f"Moving: {p_from} => {p_to}, similarity: {p_similarity}")
        move_elements_across_dir(
            p_from,
            p_to,
            replace_options=REPLACE_OPTION_UPDATE_PACK,
        )


if __name__ == "__main__":
    main()

File: src/wasted/aery_fix.rs
//! Aery fix utility.
//!
//! This module provides a utility for fixing Aery-themed BMS directories.

use std::path::{Path, PathBuf};

use crate::fs::name::bms_dir_similarity;
use crate::fs::pack_move::{MoveOptions, REPLACE_OPTION_UPDATE_PACK, move_elements_across_dir};

/// Fix Aery folders by merging similar directories
///
/// # Errors
///
/// Returns [`std::io::Error`] if directory operations fail.
///
/// # Panics
///
/// Panics if stdout flush fails.
pub fn aery_fix(src_dir: &Path) -> Result<(), std::io::Error> {
    use std::io::{self, Write};

    if !src_dir.is_dir() {
        println!("{}: not a dir.", src_dir.display());
        return Ok(());
    }

    // Get all directories
    let dirs: Vec<String> = std::fs::read_dir(src_dir)?
        .filter_map(std::result::Result::ok)
        .filter(|e| e.path().is_dir())
        .map(|e| e.file_name().to_string_lossy().into_owned())
        .collect();

    // Find all Aery directories and their matches
    let mut aery_pairs: Vec<(PathBuf, PathBuf, f64)> = Vec::new();

    for dir in &dirs {
        if !dir.contains("Aery") && !dir.contains("AERY") {
            continue;
        }

        let dir_path = src_dir.join(dir);

        // Look for matching directories by prefix
        for i in 0..dir.len() {
            let prefix = &dir[..=i];
            let matching_dirs: Vec<_> = dirs
                .iter()
                .filter(|d| d.starts_with(prefix) && *d != dir)
                .collect();

            if matching_dirs.len() == 1 {
                let scan_dir = matching_dirs[0];
                let scan_dir_path = src_dir.join(scan_dir);

                let similarity = bms_dir_similarity(&dir_path, &scan_dir_path);
                aery_pairs.push((dir_path.clone(), scan_dir_path.clone(), similarity));
                break;
            }
        }
    }

    // Print pairs
    for p in &aery_pairs {
        println!(
            "{} => {}, similarity: {}",
            p.0.display(),
            p.1.display(),
            p.2
        );
    }

    let similarity_border = 0.95;

    print!("Confirm? (border: {similarity_border}) [y/N]");
    io::stdout().flush().unwrap();

    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let input_stripped = input
        .strip_suffix('\n')
        .or_else(|| input.strip_suffix("\r\n"))
        .unwrap_or(&input);
    if !input_stripped.to_lowercase().starts_with('y') {
        return Ok(());
    }

    // Execute moves
    for (p_from, p_to, similarity) in aery_pairs {
        if similarity < similarity_border {
            continue;
        }
        println!(
            "Moving: {} => {}, similarity: {similarity}",
            p_from.display(),
            p_to.display()
        );
        move_elements_across_dir(
            &p_from,
            &p_to,
            MoveOptions::default(),
            &REPLACE_OPTION_UPDATE_PACK,
        )?;
    }

    Ok(())
}
